### PR TITLE
argc: 1.14.0 -> 1.19.0

### PR DESCRIPTION
**Pull request by Pizmovc**
## Description of changes
There were quite a few releases in between, so its best to check the [source repo releases](https://github.com/sigoden/argc/releases).

## Things tried :sweat_smile: 

When I tried building the package the tests of the package failed though, and I was not able to figure out why.

EDIT: The issue was resolved, see my comment below.

<details>
<summary>Here is the failed test result:</summary>

```
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot file: tests/snapshots/integration__misc__escape.snap
Snapshot: escape
Source: tests/misc.rs:5
────────────────────────────────────────────────────────────────────────────────
Expression: data
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
   10    10 │ cmd '$foo' '`pwd`' '$(pwd)' ''\''' '\1' '' '
   11    11 │ ' 世界 ' '
   12    12 │ 
   13    13 │ # BUILD_OUTPUT
   14       │-argc__args=([0]="prog" [1]="cmd" [2]="\$foo" [3]="\`pwd\`" [4]="\$(pwd)" [5]="'" [6]="\\1" [7]="" [8]=$'\n' [9]="世界" [10]=" ")
         14 │+argc__args=([0]="prog" [1]="cmd" [2]="\$foo" [3]="\`pwd\`" [4]="\$(pwd)" [5]="'" [6]="\\1" [7]="" [8]=$'\n' [9]=$'\344\270\226\347\225\214' [10]=" ")
   15    15 │ argc__fn=cmd
   16       │-argc__positionals=([0]="\$foo" [1]="\`pwd\`" [2]="\$(pwd)" [3]="'" [4]="\\1" [5]="" [6]=$'\n' [7]="世界" [8]=" ")
         16 │+argc__positionals=([0]="\$foo" [1]="\`pwd\`" [2]="\$(pwd)" [3]="'" [4]="\\1" [5]="" [6]=$'\n' [7]=$'\344\270\226\347\225\214' [8]=" ")
   17    17 │ cmd $foo `pwd` $(pwd) ' \1  
   18    18 │  世界
────────────┴───────────────────────────────────────────────────────────────────
```

</details>

Hopefully the tests will pass on the CI [as they do in the source repo](https://github.com/sigoden/argc/actions/runs/9563590037/job/26362403842) :crossed_fingers: 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Pizmovc** [comment]
cc @ShamrockLee this needs some help I think :pray:

**Comment by ShamrockLee** [comment]
So the problem is that we expect the Unicode associative array value be preserved as something like `"世界"`, but it turns out to be `$'\344\270\226\347\225\214'`.

 `$'\344\270\226\347\225\214'` is just the UTF-8-encoded `"世界"` in octal. Guess what,

```console
$ echo $'\344\270\226\347\225\214'
世界
```

**Comment by Pizmovc** [comment]
What can we do to push this forward?

**Comment by ShamrockLee** [comment]
> I verified that the `1.15.0` already contains the change that breaks our execution of the same tests.

The test case was changed in sigoden/argc#295, the PR that adds the `argc build` support, and the failing part is related to the `argc build`.

Considering this, it is not some changes from the upstream or us that break the test -- it fails from the beginning.

> What can we do to push this forward?

Let's post it under the [PRs in distress](https://discourse.nixos.org/t/prs-in-distress/3604) NixOS Discourse topic. I hope someone can help out after seeing that.

**Comment by Pizmovc** [comment]
@ShamrockLee thanks for investigating this further. It was valuable info for me.

I played around with this a bit more today and (with the help of Claude) I was able to find a fix!
It seems like adding these two environment variables fixes the tests.

This is what I had to add to make pass the tests.

```nix
  buildInputs = [ pkgs.glibcLocales ];

  LANG = "en_US.UTF-8";
  LOCALE_ARCHIVE = "${pkgs.glibcLocales}/lib/locale/locale-archive";
```

It make sense to me in a way, but I'm no C or Rust developer, so if there is a better way of achieving this feel free to chime in.

**Comment by ShamrockLee** [comment]
Congratulations on finding the solution to the test failure!

I have considered changing the LOCALE, but I need to familiarize myself with it.

It doesn't sound right for a compiler-like program to emit code based on the locale of the developer's machine. It's okay to set them as attributes for testing purposes anyway.

**Review by ShamrockLee — CHANGES_REQUESTED** [review]
Please pass `glibcLocales` directly instead of passing `pkgs`. Passing `pkgs` directly would make `<pkg>.override`-ing difficult.

File: pkgs/by-name/ar/argc/package.nix [hunk 1]
```diff
@@ -37,6 +38,11 @@ rustPlatform.buildRustPackage rec {
 
   disallowedReferences = lib.optional (!canExecuteHost) buildPackages.argc;
 
+  buildInputs = [ pkgs.glibcLocales ];
+
+  LANG = "en_US.UTF-8";
+  LOCALE_ARCHIVE = "${pkgs.glibcLocales}/lib/locale/locale-archive";
```
Comment (ShamrockLee) [inline pkgs/by-name/ar/argc/package.nix:44]:
```suggestion
  env.LANG = "en_US.UTF-8";
  env.LOCALE_ARCHIVE = "${glibcLocales}/lib/locale/locale-archive";
```

**Comment by rupurt** [comment]
@ByteSudoer has created a 1.19.0 update that's passing the CI build in https://github.com/NixOS/nixpkgs/pull/329189

File: pkgs/by-name/ar/argc/package.nix [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {
   lib,
   buildPackages,
+  pkgs,
   pkgsCross,
   rustPlatform,
   stdenv,
```
[the remark this answers is not in the document]
Comment (Pizmovc) [reply pkgs/by-name/ar/argc/package.nix:6]: Fixed in 8d9fa3c75d4c

**Comment by Pizmovc** [comment]
I've accepted all of you suggestions @ShamrockLee, thank you for making them!

**Comment by Pizmovc** [comment]
> @ByteSudoer has created a 1.19.0 update that's passing the CI build in #329189

Thank you for letting us know, though I would prefer if we merged this MR instead of the linked one, since we've identified the underlying issue and fixed the tests instead of skipping them.

**Comment by ShamrockLee** [comment]
nixpkgs-review runs successfully on my laptop. I wonder why the OfBorg test is failing.

**Comment by ShamrockLee** [comment]
@ofborg eval

File: pkgs/by-name/ar/argc/package.nix [hunk 3]
```diff
@@ -37,6 +38,11 @@ rustPlatform.buildRustPackage rec {
 
   disallowedReferences = lib.optional (!canExecuteHost) buildPackages.argc;
 
+  env = {
+    LANG = "en_US.UTF-8";
```
Comment (ShamrockLee) [inline pkgs/by-name/ar/argc/package.nix:42]: I found that `LANG=C.UTF-8` would also work, which sounds more language- and location-neutral.

```suggestion
    LANG = "C.UTF-8";
```

File: pkgs/by-name/ar/argc/package.nix [hunk 4]
```diff
@@ -37,6 +38,11 @@ rustPlatform.buildRustPackage rec {
...
+    LOCALE_ARCHIVE = "${glibcLocales}/lib/locale/locale-archive";
```
Comment (ShamrockLee) [inline pkgs/by-name/ar/argc/package.nix:43]: Ah, I got it!

https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/all-packages.nix#L20736-L20744


```suggestion
    LOCALE_ARCHIVE = lib.optionalString (glibcLocale != null) "${glibcLocales}/lib/locale/locale-archive";
```

Comment (Pizmovc) [reply pkgs/by-name/ar/argc/package.nix]: I was scratching my head about what do with this one :)

Great work! I've included the suggestion in 4d248dad5e5391126c3730003b469c022e041713.

File: pkgs/by-name/ar/argc/package.nix [hunk 5]
```diff
@@ -37,6 +38,13 @@ rustPlatform.buildRustPackage rec {
 
   disallowedReferences = lib.optional (!canExecuteHost) buildPackages.argc;
 
+  env = {
+    LANG = "C.UTF-8";
+    LOCALE_ARCHIVE = lib.optionalString (
+      glibcLocales != null
+    ) "${glibcLocales}/lib/locale/locale-archive";
+  };
```
Comment (SuperSandro2000) [inline pkgs/by-name/ar/argc/package.nix:46]:
```suggestion
    LANG = "C.UTF-8";
  } // lib.optionalAttrs (glibcLocales != null) {
    LOCALE_ARCHIVE = "${glibcLocales}/lib/locale/locale-archive";
  };
```
a little bit cleaner

**Comment by ShamrockLee** [comment]
@SuperSandro2000 Should we add run-time `LOCALE_ARCHIVE` with `wrapProgram`? The test failure without it seems indicating that `argc build` would only emit Unicode strings correctly if `LOCALE_ARCHIVE` is present.

File: pkgs/by-name/ar/argc/package.nix (hunk 5 above)
Comment (Pizmovc) [reply pkgs/by-name/ar/argc/package.nix]: Yeah, good point. Implemented in ac93a0467d7a9b85d62fd10e838ad362ec458c83.

**Comment by SuperSandro2000** [comment]
No we shouldn't. Otherwise we overwrite the users locale settings.

**Pull request merged**